### PR TITLE
Add type validation to non-template .arb file parsing logic

### LINKED ISSUE #138297: [tool_crash] _TypeError: (#0      AppResourceBundle.translationFor (package:flutter_tools/src/localizations/gen_l10n_types.dart:637:70))
## Command
```
flutter --no-color run --machine --track-widget-creation --device-id=OF6LHMB9A2101175 --start-paused --dart-define=flutter.inspector.structuredErrors=true lib\main.dart
```

## Steps to Reproduce
1. ...
2. ...
3. ...

## Logs
_TypeError: (#0      AppResourceBundle.translationFor (package:flutter_tools/src/localizations/gen_l10n_types.dart:637:70))
```
#0      AppResourceBundle.translationFor (package:flutter_tools/src/localizations/gen_l10n_types.dart:637:70)
#1      new Message (package:flutter_tools/src/localizations/gen_l10n_types.dart:353:42)
#2      LocalizationsGenerator.loadResources.<anonymous closure> (package:flutter_tools/src/localizations/gen_l10n.dart:908:67)
#3      MappedIterator.moveNext (dart:_internal/iterable.dart:393:20)
#4      new _GrowableList._ofOther (dart:core-patch/growable_array.dart:202:26)
#5      new _GrowableList.of (dart:core-patch/growable_array.dart:152:26)
#6      new List.of (dart:core-patch/array_patch.dart:47:28)
#7      Iterable.toList (dart:core/iterable.dart:495:7)
#8      LocalizationsGenerator.loadResources (package:flutter_tools/src/localizations/gen_l10n.dart:910:8)
#9      generateLocalizations (package:flutter_tools/src/localizations/gen_l10n.dart:73:9)
#10     GenerateLocalizationsTarget.build (package:flutter_tools/src/build_system/targets/localizations.dart:60:11)
#11     _BuildInstance._invokeInternal (package:flutter_tools/src/build_system/build_system.dart:853:27)
<asynchronous suspension>
#12     FlutterBuildSystem.build (package:flutter_tools/src/build_system/build_system.dart:620:16)
<asynchronous suspension>
#13     generateLocalizationsSyntheticPackage (package:flutter_tools/src/dart/generate_synthetic_packages.dart:56:30)
<asynchronous suspension>
#14     FlutterCommand.verifyThenRunCommand (package:flutter_tools/src/runner/flutter_command.dart:1592:7)
<asynchronous suspension>
#15     FlutterCommand.run.<anonymous closure> (package:flutter_tools/src/runner/flutter_command.dart:1297:27)
<asynchronous suspension>
#16     AppContext.run.<anonymous closure> (package:flutter_tools/src/base/context.dart:150:19)
<asynchronous suspension>
#17     CommandRunner.runCommand (package:args/command_runner.dart:212:13)
<asynchronous suspension>
```
```
[✓] Flutter (Channel stable, 3.13.8, on Microsoft Windows [Version 10.0.22621.2506], locale en-GB)
    • Flutter version 3.13.8 on channel stable at D:\Flutter\flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 6c4930c4ac (3 weeks ago), 2023-10-18 10:57:55 -0500
    • Engine revision 767d8c75e8
    • Dart version 3.1.4
    • DevTools version 2.25.0

[✓] Windows Version (Installed version of Windows is version 10 or higher)

[!] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at C:\Users\Administrator\AppData\Local\Android\sdk
    ✗ cmdline-tools component is missing
      Run `path/to/sdkmanager --install "cmdline-tools;latest"`
      See https://developer.android.com/studio/command-line for more details.
    ✗ Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/windows#android-setup for more details.

[✓] Chrome - develop for the web
    • Chrome at C:\Program Files\Google\Chrome\Application\chrome.exe

[✗] Visual Studio - develop Windows apps
    ✗ Visual Studio not installed; this is necessary to develop Windows apps.
      Download at https://visualstudio.microsoft.com/downloads/.
      Please install the "Desktop development with C++" workload, including all of its default components

[✓] Android Studio (version 2022.3)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-b2043.56-10027231)

[✓] IntelliJ IDEA Ultimate Edition (version 2023.1)
    • IntelliJ at C:\Program Files\JetBrains\IntelliJ IDEA 2023.1.2
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.84.1)
    • VS Code at C:\Users\Administrator\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.76.0

[✓] Connected device (4 available)
    • SHV46 (mobile)    • OF6LHMB9A2101175 • android-arm64  • Android 11 (API 30)
    • Windows (desktop) • windows          • windows-x64    • Microsoft Windows [Version 10.0.22621.2506]
    • Chrome (web)      • chrome           • web-javascript • Google Chrome 118.0.5993.88
    • Edge (web)        • edge             • web-javascript • Microsoft Edge 119.0.2151.58

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 2 categories.

```

## Flutter Application Metadata
**Type**: malformed
**Version**: 1.0.0+1
**Material**: true
**Android X**: false
**Module**: false
**Plugin**: false
**Android package**: null
**iOS bundle identifier**: null
**Creation channel**: null
**Creation framework version**: null
### Plugins
firebase_core-2.22.0
firebase_core_web-2.8.1
firebase_messaging-14.7.4
firebase_messaging_web-3.5.13
flutter_native_select-1.0.1
flutter_native_splash-2.2.19
path_provider_linux-2.2.1
path_provider_windows-2.2.1
restart_app-1.2.1
shared_preferences-2.1.0
shared_preferences_android-2.2.1
shared_preferences_foundation-2.3.4
shared_preferences_linux-2.3.2
shared_preferences_web-2.2.1
shared_preferences_windows-2.3.2

### PR DESCRIPTION
**Pull request by andrewkolos**
Resolves https://github.com/flutter/flutter/issues/138297. 

When reading from a .arb file (which contains localizations of strings in the form of a user-defined JSON string-string map), validate the type of values that we read.

See [this comment of mine on #138297](https://github.com/flutter/flutter/issues/138297#issuecomment-1827043260) to see how I arrived at this fix.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter_tools/test/general.shard/localizations/gen_l10n_test.dart [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+void main() {
+  late FileSystem fs;
+  late BufferLogger logger;
+
+  setUp(() {
+    fs = MemoryFileSystem.test();
+    logger = BufferLogger.test();
+  });
+
+  testWithoutContext('AppResourceBundle throws if file contains non-string value', () {
```
Comment (christopherfujino) [inline packages/flutter_tools/test/general.shard/localizations/gen_l10n_test.dart:21]: nit. could/should this be in https://github.com/flutter/flutter/blob/master/packages/flutter_tools/test/general.shard/generate_localizations_test.dart ?

**Review by christopherfujino — APPROVED** [review]
nit about which library to put the test, otherwise LGTM

File: packages/flutter_tools/test/general.shard/localizations/gen_l10n_test.dart (hunk 1 above)
Comment (andrewkolos) [reply packages/flutter_tools/test/general.shard/localizations/gen_l10n_test.dart]: I was wondering where the non-integration tests for this stuff was. Fixed.

Comment (christopherfujino) [reply packages/flutter_tools/test/general.shard/localizations/gen_l10n_test.dart]: your file name was better, FWIW

**Pull request merged**